### PR TITLE
BZ-1890452 Azure BYOL disk encryption through DES

### PR DESCRIPTION
**Pull request by lbarbeevargas**
https://bugzilla.redhat.com/show_bug.cgi?id=1890452 (dev BZ)
Also pertains to https://bugzilla.redhat.com/show_bug.cgi?id=1946959 (docs BZ)
https://issues.redhat.com/browse/OCPCLOUD-1058 documented by https://issues.redhat.com/browse/OSDOCS-2170.

Customer-managed encryption keys for a machine set - Azure. **Requires docs change management approval.**

Preview: https://deploy-preview-32330--osdocs.netlify.app/openshift-enterprise/latest/machine_management/creating_machinesets/creating-machineset-azure.html#machineset-enabling-customer-managed-encryption-azure_creating-machineset-azure

### CODE REVIEW

**Comment by lbarbeevargas** [comment]
@JoelSpeed @kwoodson PTAL. This is the first pass at customer-managed encryption keys for Azure. Also, please let me know if we should include the instructions on how to set up the key vault, or if the current link to the Azure docs works. Thank you.

**Comment by lbarbeevargas** [comment]
@JoelSpeed PTAL - this draft includes a Prerequisites section where users first set up a key vault and disk encryption set.

**Comment by JoelSpeed** [comment]
I think it's ok to link out to these Azure docs as pre-requisistes, they are clear and explain what is required before we add the docs for our own part of this.

/lgtm

**Comment by lbarbeevargas** [comment]
@vikram-redhat Can you PTAL to confirm that it is okay to link out to the Azure docs in the Prerequisites vs reproducing those steps in the OCP docs? Thank you.

**Comment by vikram-redhat** [comment]
Generally, I think this should be ok, but I am going to punt to @sferich888 for an ack.

**Comment by lbarbeevargas** [comment]
> Generally, I think this should be ok, but I am going to punt to @sferich888 for an ack.

Thank you. FWIW, I documented the [customer-managed encryption keys section for GCP](https://docs.openshift.com/container-platform/4.7/machine_management/creating_machinesets/creating-machineset-gcp.html#machineset-enabling-customer-managed-encryption_creating-machineset-gcp) in a similar manner. I plan on updating the GCP section after this PR merges to better align the modules (e.g., add a Prerequisites section and mirror the `providerSpec` example).

**Comment by vikram-redhat** [comment]
We are good to include these links @lbarbeevargas

**Comment by lbarbeevargas** [comment]
@sunzhaohua2 PTAL, thank you.

File: modules/machineset-customer-managed-encryption-azure.adoc [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// Module included in the following assemblies:
+//
+// * machine_management/creating_machinesets/creating-machineset-gcp.adoc
+
+[id="machineset-enabling-customer-managed-encryption-azure_{context}"]
+= Enabling customer-managed encryption keys for a machine set
+
+Azure allows users to supply an encryption key to encrypt data on managed disks at rest. You can enable server-side encryption with customer-managed keys by using the Machine API.
```
Comment (maxwelldb) [inline modules/machineset-customer-managed-encryption-azure.adoc:8]: ⚠️ ISG prefers user-centered language over "allows" in cases like this. E.g., "You can supply an encryption key to Azure to encrypt data on managed disks at rest."

File: modules/machineset-customer-managed-encryption-azure.adoc [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+An Azure Key Vault, a disk encryption set, and an encryption key are required to use a customer-managed key. The disk encryption set must preside in a resource group where the Cloud Credential Operator (CCO) has granted permissions. If not, an additional reader role is required to be granted on the disk encryption set.
```
Comment (maxwelldb) [inline modules/machineset-customer-managed-encryption-azure.adoc:10]: Is this paragraph essentially restating (prestating?) the prerequisites?

File: modules/machineset-customer-managed-encryption-azure.adoc [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+.Prerequisites
+
+* link:https://docs.microsoft.com/en-us/azure/aks/azure-disk-customer-managed-keys#create-an-azure-key-vault-instance[Create an Azure Key Vault instance].
+* link:https://docs.microsoft.com/en-us/azure/aks/azure-disk-customer-managed-keys#create-an-instance-of-a-diskencryptionset[Create an instance of a disk encryption set].
+* link:https://docs.microsoft.com/en-us/azure/aks/azure-disk-customer-managed-keys#grant-the-diskencryptionset-access-to-key-vault[Grant the disk encryption set access to key vault].
+
+.Procedure
+
+. Configure the disk encryption set under the `providerSpec` field in your machine set YAML file. For example:
```
Comment (maxwelldb) [inline modules/machineset-customer-managed-encryption-azure.adoc:20]: If we can expect readers to understand this schema pretty well if they're reading this module, this is cool. If they might not know it too well, we might consider adding some more info here w/callouts.

**Review by maxwelldb — APPROVED** [review]
Looks copacetic. Added a few comments for consideration.

File: modules/machineset-customer-managed-encryption-azure.adoc (hunk 2 above)
Comment (lbarbeevargas) [reply modules/machineset-customer-managed-encryption-azure.adoc:10]: Somewhat, it's also providing the info that:

> The disk encryption set must preside in a resource group where the Cloud Credential Operator (CCO) has granted permissions. If not, an additional reader role is required to be granted on the disk encryption set.

**Comment by mgahagan73** [comment]
I've tested BYOK a few times on Azure with 4.8, doc change looks good to me.

**Pull request merged**